File: Max Consecutive Ones III - Leetcode 1004/Max Consecutive Ones III - Leetcode 1004.py
class Solution:
    def longestOnes(self, nums: List[int], k: int) -> int:
        max_w = 0
        num_zeros = 0
        n = len(nums)
        l = 0

        for r in range(n):
            if nums[r] == 0:
                num_zeros += 1

            while num_zeros > k:
                if nums[l] == 0:
                    num_zeros -= 1
                l += 1
            w = r - l + 1
            max_w = max(max_w, w)

        return max_w

# Time Complexity: O(n)
# Space Complexity: O(1)

File: Max Consecutive Ones III - Leetcode 1004/Max Consecutive Ones III - Leetcode 1004.cpp
class Solution {
public:
    int longestOnes(vector<int>& nums, int k) {
        int max_w = 0;
        int num_zeros = 0;
        int l = 0;

        for (int r = 0; r < nums.size(); ++r) {
            if (nums[r] == 0) {
                num_zeros++;
            }

            while (num_zeros > k) {
                if (nums[l] == 0) {
                    num_zeros--;
                }
                l++;
            }

            max_w = max(max_w, r - l + 1);
        }

        return max_w;
    }
};
